File: server/src/services/fatigueService.js
import { redis } from '../config/redis.js';

/**
 * Using Redis ZSET for Sliding Window fatigue checks.
 */
export const checkFatigue = async (userId, category) => {
    const now = Date.now();
    const key = `fatigue:${userId}:${category}`;

    // Clean up old events (outside the 10-minute window)
    const windowStart = now - 600000; // 10 minutes
    await redis.zremrangebyscore(key, 0, windowStart);

    // Count events in the last 10 minutes
    const count = await redis.zcount(key, '-inf', '+inf');

    // Level 1: > 5 alerts in 10 mins -> Demote
    if (count >= 5) {
        return { fatigue: true, level: 1, action: 'DEMOTE_TO_LATER' };
    }

    // Level 2: check hourly (demo)
    // ...

    return { fatigue: false };
};

export const recordEventForFatigue = async (userId, category) => {
    const key = `fatigue:${userId}:${category}`;
    await redis.zadd(key, Date.now(), `${Date.now()}-${userId}`);
    await redis.expire(key, 3600); // 1 hour max retention
};

File: spring-engine/src/main/java/com/priorityengine/service/FatigueService.java
package com.priorityengine.service;

import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;
import java.time.Instant;
import java.util.Set;
import java.util.concurrent.TimeUnit;

@Service
@RequiredArgsConstructor
public class FatigueService {

    private final StringRedisTemplate redisTemplate;

    public boolean isUserFatigued(String userId, String category) {
        String key = "fatigue:" + userId + ":" + category;
        long now = Instant.now().toEpochMilli();
        long windowStart = now - 600000; // 10 minutes

        // 1. Clean up old events (Sliding Window in Redis ZSET)
        redisTemplate.opsForZSet().removeRangeByScore(key, 0, windowStart);

        // 2. Count current events
        Long count = redisTemplate.opsForZSet().count(key, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);

        // 3. Level 1: > 5 alerts in 10 mins
        if (count != null && count >= 5) {
            return true;
        }

        // 4. Record current event
        redisTemplate.opsForZSet().add(key, String.valueOf(now), now);
        redisTemplate.expire(key, 1, TimeUnit.HOURS);
        return false;
    }
}
